cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -723,53 +723,85 @@ func fixNodeGroupSize(context *context.AutoscalingContext, clusterStateRegistry
+	for _, un := range unregisteredNodes {
+		nodeGroup, err := a.CloudProvider.NodeGroupForNode(un.Node)
 		if err != nil {
-			klog.Warningf("Failed to get node group for %s: %v", unregisteredNode.Node.Name, err)
-			return removedAny, err
-		}
-		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
-			klog.Warningf("No node group for node %s, skipping", unregisteredNode.Node.Name)
+			id := "<nil>"
+			if un.Node != nil {
+				id = un.Node.Spec.ProviderID
+			}
+			klog.Warningf("Cannot determine nodeGroup for node %v; %v", id, err)
 			continue
 		}
 
 		maxNodeProvisionTime, err := csr.MaxNodeProvisionTime(nodeGroup)
 		if err != nil {
-			return removedAny, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", unregisteredNode.Node.Name, nodeGroup.Id())
+			return false, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", un.Node.Name, nodeGroup.Id())
 		}
 
-		if unregisteredNode.UnregisteredSince.Add(maxNodeProvisionTime).Before(currentTime) {
-			klog.V(0).Infof("Removing unregistered node %v", unregisteredNode.Node.Name)
-			size, err := nodeGroup.TargetSize()
-			if err != nil {
-				klog.Warningf("Failed to get node group size; unregisteredNode=%v; nodeGroup=%v; err=%v", unregisteredNode.Node.Name, nodeGroup.Id(), err)
-				continue
+		if un.UnregisteredSince.Add(maxNodeProvisionTime).Before(currentTime) {
+			if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+				a.clusterStateRegistry.RefreshCloudProviderNodeInstancesCache()
+				return false, fmt.Errorf("node %s has no known nodegroup", un.Node.GetName())
 			}
-			if nodeGroup.MinSize() >= size {
-				klog.Warningf("Failed to remove node %s: node group min size reached, skipping unregistered node removal", unregisteredNode.Node.Name)
+			nodesToBeDeletedByNodeGroupId[nodeGroup.Id()] = append(nodesToBeDeletedByNodeGroupId[nodeGroup.Id()], un)
+		}
+	}
+
+	removedAny := false
+	for nodeGroupId, nodesToDelete := range nodesToBeDeletedByNodeGroupId {
+		nodeGroup := nodeGroups[nodeGroupId]
+
+		klog.V(0).Infof("Removing %v unregistered nodes for node group %v", len(nodesToDelete), nodeGroupId)
+		size, err := nodeGroup.TargetSize()
+		if err != nil {
+			klog.Warningf("Failed to get node group size; nodeGroup=%v; err=%v", nodeGroup.Id(), err)
+			continue
+		}
+		roomToDelete := size - nodeGroup.MinSize()
+		if roomToDelete < 0 {
+			roomToDelete = 0
+		}
+		if len(nodesToDelete) > roomToDelete {
+			klog.Warningf("Node group %s min size reached, removing only %v out of %v unregistered nodes", nodeGroupId, len(nodesToDelete), roomToDelete)
+			nodesToDelete = nodesToDelete[:roomToDelete]
+			if roomToDelete == 0 {
 				continue
 			}
-			err = nodeGroup.DeleteNodes([]*apiv1.Node{unregisteredNode.Node})
-			csr.InvalidateNodeInstancesCacheEntry(nodeGroup)
-			if err != nil {
-				klog.Warningf("Failed to remove node %s: %v", unregisteredNode.Node.Name, err)
+		}
```

Comment:
Yes, this is also implied by the name of this option. You're right though, we might be missing some logic to enforce that. I could file a follow-up bug on this, WDYT?

Reply:
Yes, I think it would be good to do. In my opinion we need to either have some sort of enforcement option like "if `ZeroOrMaxNodeScaling` then we override the min size to 0", or some validation like "node group needs to have both `ZeroOrMaxNodeScaling` and min size set to 0 for the logic to be triggered".

Reply:
Agreed, will file a follow-up bug to do that.